[Stdlib] Add implicit Codepoint(StringLiteral) constructor

KrxGu opened the pull request:
Fixes #5490 

This adds an implicit constructor to `Codepoint` that accepts a compile-time `StringLiteral` parameter, making it much easier to create codepoint constants.

### What changed
- Added `__init__[value: __mlir_type.\`!kgen.string\`](out self, lit: StringLiteral[value])`
- Constructor validates at compile-time that the literal is exactly 1 byte
- Computes the codepoint value using `alias` for compile-time evaluation
- Added comprehensive tests covering ASCII, whitespace, and compile-time usage

### Why
Currently, the only way to create a `Codepoint` from a literal is `Codepoint.ord("x")`, which is verbose and can generate suboptimal IR. The stdlib has been working around this with patterns like:
```mojo
alias ` ` = UInt8(ord(" "))
alias `\t` = UInt8(ord("\t"))
```

Now you can just write:
```
var space = Codepoint(" ")
var tab = Codepoint("\t")
```

The constructor enforces the single-byte constraint at compile-time, so invalid literals like Codepoint("") or Codepoint("ab") will fail to build rather than causing runtime issues.

### Testing
./bazelw test //mojo/stdlib/test/collections:test_codepoint.mojo.test passes
Added test_char_from_stringliteral() with coverage for ASCII, whitespace, and compile-time contexts

martinvuyk commented on mojo/stdlib/stdlib/collections/string/codepoint.mojo:
````diff
@@ -138,6 +138,56 @@ struct Codepoint(
+        ```
+        """
+        # Reconstruct the literal from the type parameter to force compile-time evaluation
+        alias sl: StringLiteral[value] = {}
````
Nice trick

martinvuyk replied:
```suggestion
        comptime sl: StringLiteral[value] = {}
```

martinvuyk commented on mojo/stdlib/stdlib/collections/string/codepoint.mojo:
```diff
@@ -138,6 +138,56 @@ struct Codepoint(
+        # Compile-time constraint - fails if literal isn't exactly 1 byte
+        constrained[
+            len(sl) == 1, "StringLiteral must contain exactly one byte"
+        ]()
```
`StringLiteral.__len__` (to my chagrin) doesn't yet work by unicode codepoints but by bytes. @JoeLoser I really want to fix #5281 before 1.0, not having strings work by unicode codepoints leads to really subtle bugs

```suggestion
        __comptime_assert sl.byte_length() > 0, "Cannot construct an empty codepoint"
        __comptime_assert sl.byte_length() == _utf8_first_byte_sequence_length(sl.as_bytes()[0]), "StringLiteral must contain exactly one codepoint"
```

martinvuyk commented on mojo/stdlib/stdlib/collections/string/codepoint.mojo:
```diff
@@ -138,6 +138,56 @@ struct Codepoint(
+        alias cp = Codepoint.ord(StaticString(sl))
+
+        # Assign the computed value
+        self = cp
```
```suggestion
        comptime cp = Codepoint.ord(s)) # shift compute cost to compile-time
        self = cp
```

martinvuyk reviewed (commented):
Hi, thanks for the PR :)

martinvuyk reviewed (commented):
LGTM. Now let's wait for someone in the team to look this over and approve it

Side-note: Please only ping whomever was assigned to your PR (especially avoid pinging Joe without need). I have no problem in helping out

NathanSWard commented on mojo/stdlib/stdlib/collections/string/codepoint.mojo:
```diff
@@ -138,6 +138,68 @@ struct Codepoint(
+        # Compile-time validation for proper UTF-8 codepoint
+        constrained[
+            sl.byte_length() > 0, "Cannot construct an empty codepoint"
+        ]()
```
In the stdlib - we're moving away from `constrained` and now using `__comptime_assert` :) 

```suggestion
        __comptime_assert  sl.byte_length() > 0, "Cannot construct an empty codepoint"
```

NathanSWard commented on mojo/stdlib/stdlib/collections/string/codepoint.mojo:
```diff
@@ -138,6 +138,68 @@ struct Codepoint(
+            sl.as_bytes()
+        )
+
+        constrained[
```
Ditto `__comptime_assert`

ConnorGray commented on mojo/stdlib/stdlib/collections/string/codepoint.mojo:
```diff
@@ -138,6 +138,67 @@ struct Codepoint(
+    @implicit
+    fn __init__[
+        value: __mlir_type.`!kgen.string`
+    ](out self, lit: StringLiteral[value]):
```
**Suggestion** I suggest we remove `@implicit` unless there is a very strong reason to keep it. Implicit conversions have a way of popping up in unexpected places. For a "power user" feature like `Codepoint`, I think it's likely the readability advantage of making it explicit outweighs the costs of having to type it out explicitly.

**Question** Could we drop the explicit `__mlir_type.!kgen.string` parameter? This should implicitly autoparameterize, and we can still access the raw string down in the body later using `lit.value`.

```suggestion
    fn __init__(out self, lit: StringLiteral):
```

ConnorGray commented on mojo/stdlib/stdlib/collections/string/codepoint.mojo:
````diff
@@ -138,6 +138,67 @@ struct Codepoint(
+        ```
+        """
+        # Reconstruct the literal from the type parameter to force compile-time evaluation
+        comptime sl: StringLiteral[value] = {}
````
**Praise** Neat trick!

Re. my above comment, I think this could be written:


```suggestion
        comptime sl: StringLiteral[lit.value] = {}
```

but if that doesn't work, this should:

```suggestion
        comptime sl: StringLiteral[type_of(lit).value]
```

KrxGu committed:
style: apply pre-commit formatting

KrxGu committed:
Remove @implicit and simplify constructor signature

KrxGu committed:
Use __comptime_assert instead of constrained

KrxGu committed:
Use unsafe_decode_utf8_codepoint for validation consistency

KrxGu committed:
style: apply pre-commit formatting

Signed-off-by: Krish Gupta <krishgupta@Krishs-MacBook-Air.local>

KrxGu committed:
[Stdlib] Add implicit Codepoint(StringLiteral) constructor

BEGIN_PUBLIC
[Stdlib] Add implicit Codepoint(StringLiteral) constructor

Adds an implicit constructor that accepts a compile-time StringLiteral
with exactly one byte. This eliminates verbose Codepoint.ord() calls
and enables compile-time constant folding for better codegen.

Fixes #5490
END_PUBLIC

Signed-off-by: Krish Gupta <krishgupta@Krishs-MacBook-Air.local>

KrxGu committed:
Apply maintainer review suggestions - support multi-byte UTF-8

- Use 'comptime' instead of 'alias' for modern Mojo idioms
- Improve validation using _utf8_first_byte_sequence_length() to support full Unicode
- Simplify codepoint computation at compile-time
- Add test for multi-byte UTF-8 character (emoji)
